### PR TITLE
Review feedback for IApplicationLifetimeEvents

### PR DESCRIPTION
**Pull request by davidfowl**
- Renamed the type to IHostedService and added Start and Stop.
- Split up the IHostedService execution and IApplicationLifetime to avoid
circular references
- Trigger IHostedService.Start after starting the server
- Trigger IHostedService.Stop before disposing the service provider

#895 #894 

/cc @Tratcher @pakrym

### CODE REVIEW

**Comment by benaadams** [comment]
Hmm... `Start` and `Stop` sound like commands rather than things happening.

Present participle better: `Starting` and `Stopping`?

File: src/Microsoft.AspNetCore.Hosting/Internal/HostedServiceExecutor.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+                    }
+
+                    exceptions.Add(ex);
+                }
+            }
+
+            // Throw an aggregate exception if there were any exceptions
+            if (exceptions != null)
+            {
+                throw new AggregateException(exceptions);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Hosting/Internal/HostedServiceExecutor.cs]: Sure. I don't think it matters though.

File: src/Microsoft.AspNetCore.Hosting/Internal/HostedServiceExecutor.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            }
+            catch (Exception ex)
+            {
+                _logger.ApplicationError(LoggerEventIds.HostedServiceStopException, "An error occurred stopping the application", ex);
+            }
+        }
+
+        private void Execute(Action<IHostedService> callback)
+        {
+            List<Exception> exceptions = null;
```
Comment (mikaelm12) [inline src/Microsoft.AspNetCore.Hosting/Internal/HostedServiceExecutor.cs]: Would lazy initialization work better here?

File: test/Microsoft.AspNetCore.Hosting.Tests/WebHostTests.cs [hunk 3]
```diff
@@ -340,31 +337,96 @@ public void WebHostStopCallbacksDontMultipleTimes()
...
+                   .UseServer(this)
+                   .ConfigureServices(services =>
+                   {
+                       services.AddSingleton<IHostedService, TestHostedService>();
+                   })
+                   .Build();
+            var lifetime = host.Services.GetRequiredService<IApplicationLifetime>();
             lifetime.StopApplication();
+
+
```
Comment (mikaelm12) [inline test/Microsoft.AspNetCore.Hosting.Tests/WebHostTests.cs]: nit: Remove extra new line

File: src/Microsoft.AspNetCore.Hosting/Internal/HostedServiceExecutor.cs [hunk 4]
```diff
@@ -0,0 +1,70 @@
+﻿using System;
```
Comment (mikaelm12) [inline src/Microsoft.AspNetCore.Hosting/Internal/HostedServiceExecutor.cs]: Copyright and license header here

File: src/Microsoft.AspNetCore.Hosting/Internal/HostedServiceExecutor.cs (hunk 2 above)
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Hosting/Internal/HostedServiceExecutor.cs]: Happens once per app, so it doesn't matter and doesn't need to be thread safe.

File: src/Microsoft.AspNetCore.Hosting/Internal/HostedServiceExecutor.cs (hunk 4 above)
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Hosting/Internal/HostedServiceExecutor.cs]: Good catch

**Pull request closed without merging**